### PR TITLE
WINDUP-409  Allow rules metadata to annotations - @Rules( phase = ..., ....

### PR DESCRIPTION
**Pull request by OndraZizka**
.....) WIP

### CODE REVIEW

**Comment by windupgithubbot1** [comment]
Merged build triggered.

**Comment by lincolnthree** [comment]
Please incorporate the feedback provided last month and rebase. Thank you!

**Comment by OndraZizka** [comment]
Right after 2.0 release.

File: config/tests/src/test/java/org/jboss/windup/config/Insider.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import org.ocpsoft.rewrite.context.EvaluationContext;
+
+/**
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public interface Insider
+{
+    public void call(GraphRewrite event, EvaluationContext evCtx);
+}
```
Comment (OndraZizka) [inline config/tests/src/test/java/org/jboss/windup/config/Insider.java]: (18:39:25) lincolnthree1: ozizka1: it's used only once, it's in a package that is not general and clear that it's intended for more general purpose
(18:39:50) lincolnthree1: ozizka1: and in fact, this type is useful as a general utility class overall.
(18:39:51) ozizka-FN: yes. Only once - but could be used multiple times if team finds it nice
(18:40:14) ozizka-FN: Probably.
(18:40:22) lincolnthree1: ozizka1: it should be in the API
(18:40:30) lincolnthree1: ozizka1: close to WindupRuleProvider
(18:40:37) lincolnthree1: ozizka1: and IteratingRuleProvider

File: config/tests/src/test/java/org/jboss/windup/config/Insider.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package org.jboss.windup.config;
+
+import org.ocpsoft.rewrite.context.EvaluationContext;
+
+/**
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public interface Insider
```
Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/config/Insider.java]: Delete this class.

File: config/tests/src/test/java/org/jboss/windup/config/TestInsiderRuleProvider.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.jboss.windup.graph.GraphContext;
+import org.ocpsoft.rewrite.config.Configuration;
+import org.ocpsoft.rewrite.config.ConfigurationBuilder;
+import org.ocpsoft.rewrite.context.EvaluationContext;
+
+/**
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public abstract class TestInsiderRuleProvider extends WindupRuleProvider implements Insider {
```
Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/config/TestInsiderRuleProvider.java]: This class should extend from both GraphCondition and GraphOperation, and should call both evaluate() and perform() methods like you've done below.

GraphOperation and GraphCondition are abstract classes, so you'll need to add an interface that defines the same methods that they each contain. 

Rename GraphCondition and GraphOperation to AbstractGraphCondition and AbstractGraphOperation
Create the interfaces named GraphCondition and GraphOperation.

Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/config/TestInsiderRuleProvider.java]: Note. This will be a big refactoring and will impact many files. So be careful :)

Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/config/TestInsiderRuleProvider.java]: Actually, I think a simpler way is just to leave GraphCondition and GraphOperation, and but rather re-implement the functionality of each within this class. Then you don't need to worry about new interfaces or refactoring.

File: config/api/src/main/java/org/jboss/windup/config/WindupRuleProvider.java [hunk 4]
```diff
@@ -49,72 +56,103 @@ public String getID()
...
+        return ann.phase();
     }
 
     /**
-     * Specify additional meta-data about the {@link Rule} instances originating from this {@link WindupRuleProvider}.
+     * Specify additional meta-data about the {@link Rules} instances originating from this {@link WindupRuleProvider}.
      */
     public void enhanceMetadata(Context context)
     {
+        Rules ann = this.getClass().getAnnotation(Rules.class);
```
Comment (OndraZizka) [inline config/api/src/main/java/org/jboss/windup/config/WindupRuleProvider.java]: Done.

File: config/tests/src/test/java/org/jboss/windup/config/QueryConditionTest.java [hunk 5]
```diff
@@ -42,23 +44,25 @@
...
-                                TestSomeModel.class,
-                                TestWindupConfigurationExampleRuleProvider.class)
-                    .addAsAddonDependencies(
-                                AddonDependencyEntry.create("org.jboss.windup.utils:utils"),
-                                AddonDependencyEntry.create("org.jboss.windup.config:windup-config"),
-                                AddonDependencyEntry.create("org.jboss.windup.graph:windup-graph"),
-                                AddonDependencyEntry.create("org.jboss.windup.rules.apps:rules-java"),
-                                AddonDependencyEntry.create("org.jboss.forge.furnace.container:cdi")
-                    );
+            .addBeansXML()
```
Comment (OndraZizka) [inline config/tests/src/test/java/org/jboss/windup/config/QueryConditionTest.java]: Done.

File: config/api/src/main/java/org/jboss/windup/config/WindupRuleProvider.java [hunk 6]
```diff
@@ -11,17 +11,18 @@
...
 import org.jboss.windup.config.metadata.RuleMetadata;
 import org.jboss.windup.graph.GraphContext;
 import org.ocpsoft.rewrite.config.ConfigurationProvider;
-import org.ocpsoft.rewrite.config.Rule;
 import org.ocpsoft.rewrite.context.Context;
 
 
 /**
- * {@link WindupRuleProvider} provides metadata, and a list of {@link Rule} objects that are then evaluated by the
+ * {@link WindupRuleProvider} provides metadata, and a list of {@link Rules} objects that are then evaluated by the
```
Comment (OndraZizka) [inline config/api/src/main/java/org/jboss/windup/config/WindupRuleProvider.java]: Right. I had some mind eclipse I guess, or auto replace error.

File: config/tests/src/test/java/org/jboss/windup/config/TestInsiderRuleProvider.java (hunk 3 above)
Comment (OndraZizka) [inline config/tests/src/test/java/org/jboss/windup/config/TestInsiderRuleProvider.java]: Maybe that should be a different PR? I.e. leave this metadata as is, and do the refactoring in a separate PR, affecting also these metadata tests?

Comment (OndraZizka) [inline config/tests/src/test/java/org/jboss/windup/config/TestInsiderRuleProvider.java]: In general I agree that the concept can be lifted to SimpleActionRuleProvider.

This would be an interface whose implementation will simplify implementing a simple rule with one condition and one (abstract) operation, correct?

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://10.3.9.192//job/windup-pr-builder/460/
Test PASSed.

**Comment by OndraZizka** [comment]
Reworked, rebased.

**Comment by OndraZizka** [comment]
A test does not pass due to categories not working, which I'd like to discuss. But pls review the rest.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://10.3.9.192//job/windup-pr-builder/495/
Test FAILed.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://10.3.9.192//job/windup-pr-builder/502/
Test FAILed.

File: config/api/src/main/java/org/jboss/windup/config/metadata/Rules.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+     */
+    String id() default "";
+
+    /**
+     * Return the {@link RulePhase} in which the rules from this provider should be executed.
+     * <p>
+     * The default phase is {@link RulePhase#MIGRATION_RULES}.
+     */
+    //RulePhase phase() default WindupRuleProvider.DEFAULT_PHASE;
+    Class<? extends RulePhase> phase() default Implicit.class;
```
Comment (jsight) [inline config/api/src/main/java/org/jboss/windup/config/metadata/Rules.java]: The commented line should go away, and this should be set to "WindupRuleProvider#DEFAULT_PHASE". Also, the code comment needs to be updated.

File: config/api/src/main/java/org/jboss/windup/config/WindupRuleProvider.java [hunk 8]
```diff
@@ -52,30 +63,52 @@ public String getID()
...
+        if(rulesAnnotation == null)
+            return DEFAULT_PHASE;
+        if(rulesAnnotation.phase() == null)
+            return DEFAULT_PHASE;
+        return rulesAnnotation.phase();
     }
 
     /**
-     * Specify additional meta-data about the {@link Rule} instances originating from this {@link WindupRuleProvider}.
+     * Specify additional meta-data about the {@link Rule}s instances originating from this {@link WindupRuleProvider}.
```
Comment (jsight) [inline config/api/src/main/java/org/jboss/windup/config/WindupRuleProvider.java]: Either the "s" after rules should go away or the word "instances" should go away.

File: config/tests/nbactions.xml [hunk 9]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jsight) [inline config/tests/nbactions.xml]: This file should not be here.

**Comment by jsight** [comment]
I have added a few notes. I will leave the rest of the review to Lincoln, as there were a few things that I was unsure about (based upon prior conversations).

**Comment by jsight** [comment]
retest

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://10.3.9.192//job/windup-pr-builder/508/<h2>Failed Tests: <span class='status-failure'>17</span></h2><h3><a name='org.jboss.windup:ui' /><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup$ui/testReport'>org.jboss.windup:ui</a>: <span class='status-failure'>5</span></h3><ul><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup$ui/testReport/org.jboss.windup.addon.ui/WindupCommandTest/testUserRulesDirMigration/'><strong>org.jboss.windup.addon.ui.WindupCommandTest.testUserRulesDirMigration</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup$ui/testReport/org.jboss.windup.addon.ui/WindupCommandTest/testDuplicateUserRulesDirMigration/'><strong>org.jboss.windup.addon.ui.WindupCommandTest.testDuplicateUserRulesDirMigration</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup$ui/testReport/org.jboss.windup.addon.ui/WindupCommandTest/testNewMigration/'><strong>org.jboss.windup.addon.ui.WindupCommandTest.testNewMigration</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup$ui/testReport/org.jboss.windup.addon.ui/WindupCommandTest/testUserIgnoreDirMigration/'><strong>org.jboss.windup.addon.ui.WindupCommandTest.testUserIgnoreDirMigration</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup$ui/testReport/org.jboss.windup.addon.ui/WindupCommandTest/testOutputDefaultValue/'><strong>org.jboss.windup.addon.ui.WindupCommandTest.testOutputDefaultValue</strong></a></li></ul><h3><a name='org.jboss.windup.rules.apps:rules-java' /><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-java/testReport'>org.jboss.windup.rules.apps:rules-java</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-java/testReport/org.jboss.windup.rules.java.handlers/JavaClassHandlerTest/testJavaClassCondition/'><strong>org.jboss.windup.rules.java.handlers.JavaClassHandlerTest.testJavaClassCondition</strong></a></li></ul><h3><a name='org.jboss.windup.rules.apps:rules-java-ee' /><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-java-ee/testReport'>org.jboss.windup.rules.apps:rules-java-ee</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-java-ee/testReport/org.jboss.windup.rules.apps.javaee.tests/DiscoverEjbConfigurationTest/testEJBMetadataExtraction/'><strong>org.jboss.windup.rules.apps.javaee.tests.DiscoverEjbConfigurationTest.testEJBMetadataExtraction</strong></a></li></ul><h3><a name='org.jboss.windup.rules.apps:rules-java-project' /><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-java-project/testReport'>org.jboss.windup.rules.apps:rules-java-project</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-java-project/testReport/org.jboss.windup.project.operation.test/OverviewReportLineTest/testHintAndClassificationOperation/'><strong>org.jboss.windup.project.operation.test.OverviewReportLineTest.testHintAndClassificationOperation</strong></a></li></ul><h3><a name='org.jboss.windup.rules.apps:rules-xml' /><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport'>org.jboss.windup.rules.apps:rules-xml</a>: <span class='status-failure'>9</span></h3><ul><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.apps.xml.groovy/GroovyExtensionXmlRulesTest/testHintsAndClassificationOperation/'><strong>org.jboss.windup.rules.apps.xml.groovy.GroovyExtensionXmlRulesTest.testHintsAndClassificationOperation</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.xml/XMLFileNestedConditionTest/testNestedCondition/'><strong>org.jboss.windup.rules.xml.XMLFileNestedConditionTest.testNestedCondition</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.xml/XMLHintsClassificationsTest/testHintAndClassificationOperation/'><strong>org.jboss.windup.rules.xml.XMLHintsClassificationsTest.testHintAndClassificationOperation</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.xml/XMLTransformationTest/testXSLTTransformation/'><strong>org.jboss.windup.rules.xml.XMLTransformationTest.testXSLTTransformation</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.xml/XMLTransformationXMLRulesTest/testXSLTTransformation/'><strong>org.jboss.windup.rules.xml.XMLTransformationXMLRulesTest.testXSLTTransformation</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.xml/XmlFileParameterizedTest/testXmlParams/'><strong>org.jboss.windup.rules.xml.XmlFileParameterizedTest.testXmlParams</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.xml/XmlFileParameterizedTest/testXmlParamsDangling/'><strong>org.jboss.windup.rules.xml.XmlFileParameterizedTest.testXmlParamsDangling</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.xml/XmlFileParameterizedTest/testXmlParamsMultiCondition/'><strong>org.jboss.windup.rules.xml.XmlFileParameterizedTest.testXmlParamsMultiCondition</strong></a></li><li><a href='https://windup-jenkins-os1.usersys.redhat.com/job/windup-pr-builder/508/org.jboss.windup.rules.apps$rules-xml/testReport/org.jboss.windup.rules.xml.legacy/XmlJonasConfigTest/testHintsAndClassificationOperation/'><strong>org.jboss.windup.rules.xml.legacy.XmlJonasConfigTest.testHintsAndClassificationOperation</strong></a></li></ul>
Test FAILed.

File: config/tests/src/test/java/org/jboss/windup/config/test/metadata/MetadataAnnotationExecTest.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+{
+    private static final Logger log = Logging.get(MetadataAnnotationExecTest.class);
+
+    @Deployment
+    @Dependencies({
+        @AddonDependency(name = "org.jboss.forge.furnace.container:cdi"),
+        @AddonDependency(name = "org.jboss.windup.utils:utils"),
+        @AddonDependency(name = "org.jboss.windup.config:windup-config"),
+        @AddonDependency(name = "org.jboss.windup.rules.apps:rules-base"),
+        @AddonDependency(name = "org.jboss.windup.rules.apps:rules-java"),
```
Comment (jsight) [inline config/tests/src/test/java/org/jboss/windup/config/test/metadata/MetadataAnnotationExecTest.java]: This code does not appear to be using rules-base and rules-java. These can be removed.

File: config/tests/src/test/java/org/jboss/windup/config/test/metadata/MetadataAnnotationTest.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    private GraphContextFactory grCtxFactory;
+
+
+    @Deployment
+    @Dependencies({
+                @AddonDependency(name = "org.jboss.forge.furnace.container:cdi"),
+                @AddonDependency(name = "org.jboss.windup.utils:utils"),
+                @AddonDependency(name = "org.jboss.windup.config:windup-config"),
+                @AddonDependency(name = "org.jboss.windup.graph:windup-graph"),
+                @AddonDependency(name = "org.jboss.windup.rules.apps:rules-base"),
```
Comment (jsight) [inline config/tests/src/test/java/org/jboss/windup/config/test/metadata/MetadataAnnotationTest.java]: This code is not using rules-base and rules-java. These can be removed.

File: config/tests/pom.xml [hunk 12]
```diff
@@ -32,6 +32,13 @@
         <!-- Rules Dependencies -->
         <dependency>
             <groupId>org.jboss.windup.rules.apps</groupId>
+            <artifactId>rules-base</artifactId>
```
Comment (jsight) [inline config/tests/pom.xml]: I believe that this can also be removed once the dependency lines are removed from the unit test.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://10.3.8.5//job/windup-pr-builder/571/
Test FAILed.

**Comment by OndraZizka** [comment]
Retest this please

**Comment by OndraZizka** [comment]
Retest

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://10.3.8.5//job/windup-pr-builder/572/
Test FAILed.

**Comment by OndraZizka** [comment]
retest

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://10.3.8.5//job/windup-pr-builder/577/
Test FAILed.

File: config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/AndFilter.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+import org.jboss.forge.furnace.util.Predicate;
+import org.jboss.windup.config.WindupRuleProvider;
+
+/**
+ * AND predicate which needs all of the given predicates to accept.
+ * It will stop on first false if you setStopWhenKnown(true).
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public class AndFilter implements RuleFilter
```
Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/AndFilter.java]: Please stop including these filters in PRs. They are not used.

File: config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/NotFilter.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.jboss.forge.furnace.util.Predicate;
+import org.jboss.windup.config.WindupRuleProvider;
+
+/**
+ * NOT predicate which negates the result of given predicate..
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public class NotFilter implements RuleFilter
```
Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/NotFilter.java]: Please stop including these filters in PRs. They are not used.

File: config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/OrFilter.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+import org.jboss.forge.furnace.util.Predicate;
+import org.jboss.windup.config.WindupRuleProvider;
+
+/**
+ * OR predicate which needs at least one of the contained predicates to accept.
+ * It will stop on first true if you setStopWhenKnown(true).
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public class OrFilter extends AndFilter
```
Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/OrFilter.java]: Please stop including these filters in PRs. They are not used.

File: config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/PackageRulesFilter.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.HashSet;
+import java.util.Set;
+import org.jboss.windup.config.WindupRuleProvider;
+
+/**
+ * A convenient filter for rule providers enumerated as constructor params.
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public class PackageRulesFilter implements RuleFilter
```
Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/PackageRulesFilter.java]: Please stop including these filters in PRs. They are not used.

File: config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/PhaseRulesFilter.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+import org.jboss.windup.config.phase.ReportGenerationPhase;
+import org.jboss.windup.config.phase.ReportRenderingPhase;
+import org.jboss.windup.config.phase.RulePhase;
+
+/**
+ * Filters the rules with given phases.
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public class PhaseRulesFilter implements RuleFilter
```
Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/PhaseRulesFilter.java]: Please stop including these filters in PRs. They are not used.

File: config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/RuleFilter.java [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+import org.jboss.forge.furnace.util.Predicate;
+import org.jboss.windup.config.WindupRuleProvider;
+
+
+/**
+ * Convenience type with better name.
+ *
+ * @author Ondrej Zizka, ozizka at redhat.com
+ */
+public interface RuleFilter extends Predicate<WindupRuleProvider>
```
Comment (lincolnthree) [inline config/tests/src/test/java/org/jboss/windup/testutils/rulefilters/RuleFilter.java]: Please stop including these filters in PRs. They are not used.

**Comment by lincolnthree** [comment]
Incorporated by #493
